### PR TITLE
DECA4-16: [SDK] Magic Link - provide Blade JS SDK with recovery option

### PR DESCRIPTION
**Pull request by Dubiy**

### CODE REVIEW

File: src/signers/hedera/provider.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+    getTransactionReceipt(transactionId: any) {
+        return new TransactionReceiptQuery()
+            .setTransactionId(transactionId)
+            .execute(this._client);
+    }
+
+    waitForReceipt(response: any) {
+        return new TransactionReceiptQuery()
```
Comment (Volind) [inline src/signers/hedera/provider.ts:55]: Shouldn't we add logic that will be waiting for receipt?

File: src/signers/magic/MagicProvider.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    getTransactionReceipt(transactionId: any) {
+        return new TransactionReceiptQuery()
+            .setTransactionId(transactionId)
+            .execute(this._client);
+    }
+
+    waitForReceipt(response: any) {
+        return new TransactionReceiptQuery()
```
Comment (Volind) [inline src/signers/magic/MagicProvider.ts:60]: Same about waiting here.

File: src/signers/magic/MagicProvider.ts
Comment (Volind) [inline src/signers/magic/MagicProvider.ts:1]: It's almost identical to Hedera ones.

File: test/src/index.test.ts [hunk 3]
```diff
@@ -123,6 +115,15 @@ test('bladeSdk.getBalance', async () => {
     expect(result.data).toHaveProperty("tokens");
     expect(Array.isArray(result.data.tokens)).toEqual(true);
 
+    result = await bladeSdk.setUser(AccountProvider.Hedera, accountId, privateKey, completionKey);
+    checkResult(result);
+    result = await bladeSdk.getBalance("", completionKey);
+    checkResult(result);
+
+    expect(result.data).toHaveProperty("hbars");
+    expect(result.data).toHaveProperty("tokens");
```
Comment (Volind) [inline test/src/index.test.ts:124]: This part is deprecated I guess.

File: src/BladeSDK.ts [hunk 4]
```diff
@@ -141,27 +152,30 @@ export class BladeSDK {
...
         this.sdkVersion = sdkVersion;
 
         initApiService(apiKey, dAppCode, sdkEnvironment, sdkVersion, this.network, visitorId);
-        try {
-            visitorId = await decrypt(localStorage.getItem("BladeSDK.visitorId") || "", this.apiKey);
-        } catch (e) {
-            // console.log("failed to decrypt visitor id", e);
+        if (!visitorId) {
+            try {
+                visitorId = await decrypt(localStorage.getItem("BladeSDK.visitorId") || "", this.apiKey);
```
Comment (Volind) [inline src/BladeSDK.ts]: Are you reassigning argument here?

File: src/BladeSDK.ts [hunk 5]
```diff
@@ -1391,6 +1512,28 @@ export class BladeSDK {
         }
     }
 
+    private initMagic() {
+        const magicLinkPublicKey = "pk_live_46696EF18CA8E36C";
```
Comment (Volind) [inline src/BladeSDK.ts]: Should it be here?

File: src/BladeSDK.ts (hunk 4 above)
Comment (Dubiy) [reply src/BladeSDK.ts]: only if visitorId is empty

Comment (Volind) [reply src/BladeSDK.ts]: It's good that you are not changing properties of an object argument, but as a good practice you should still avoid of doing so.

File: src/signers/hedera/provider.ts (hunk 1 above)
Comment (Dubiy) [reply src/signers/hedera/provider.ts:55]: working just fine, it's polling network until receipt ready

File: src/signers/magic/MagicProvider.ts (hunk 2 above)
Comment (Dubiy) [reply src/signers/magic/MagicProvider.ts:60]: working just fine, it's polling network until receipt ready

File: src/signers/magic/MagicProvider.ts
Comment (Dubiy) [reply src/signers/magic/MagicProvider.ts:1]: yes. inspired by https://github.com/magiclabs/example-hedera/blob/master/src/MagicProvider.js

File: test/src/index.test.ts (hunk 3 above)
Comment (Dubiy) [reply test/src/index.test.ts:124]: logic is new now, but interface saved

File: src/BladeSDK.ts (hunk 5 above)
Comment (Dubiy) [reply src/BladeSDK.ts]: oops)

File: src/BladeSDK.ts (hunk 4 above)
Comment (Dubiy) [reply src/BladeSDK.ts]: fixed

**Review by Volind — APPROVED** [review]

**Pull request merged**